pkg/api/datasources_test.go

```diff
@@ -56,3 +57,107 @@ func TestDataSourcesProxy(t *testing.T) {
 		})
 	})
 }
+
+// Adding data sources with invalid URLs should lead to an error.
+func TestAddDataSource_InvalidURL(t *testing.T) {
+	defer bus.ClearBusHandlers()
+
+	sc := setupScenarioContext("/api/datasources")
+	// TODO: Make this an argument to setupScenarioContext
+	sc.t = t
+
+	sc.m.Post(sc.url, Wrap(func(c *models.ReqContext) Response {
+		return AddDataSource(c, models.AddDataSourceCommand{
+			Name: "Test",
+			Url:  "invalid:url",
+		})
+	}))
+
+	sc.fakeReqWithParams("POST", sc.url, map[string]string{}).exec()
+
+	assert.Equal(t, 400, sc.resp.Code)
+}
+
+// Adding data sources with URLs not specifying protocol should work.
+func TestAddDataSource_URLWithoutProtocol(t *testing.T) {
+	defer bus.ClearBusHandlers()
+
+	const name = "Test"
+	const url = "localhost:5432"
+
+	// Stub handler
+	bus.AddHandler("sql", func(cmd *models.AddDataSourceCommand) error {
+		assert.Equal(t, name, cmd.Name)
+		assert.Equal(t, url, cmd.Url)
+
+		cmd.Result = &models.DataSource{}
+		return nil
+	})
+
+	sc := setupScenarioContext("/api/datasources")
+	// TODO: Make this an argument to setupScenarioContext
+	sc.t = t
+
+	sc.m.Post(sc.url, Wrap(func(c *models.ReqContext) Response {
+		return AddDataSource(c, models.AddDataSourceCommand{
+			Name: name,
+			Url:  url,
+		})
+	}))
+
+	sc.fakeReqWithParams("POST", sc.url, map[string]string{}).exec()
+
+	assert.Equal(t, 200, sc.resp.Code)
+}
+
+// Updating data sources with invalid URLs should lead to an error.
+func TestUpdateDataSource_InvalidURL(t *testing.T) {
+	defer bus.ClearBusHandlers()
+
+	sc := setupScenarioContext("/api/datasources/1234")
+	// TODO: Make this an argument to setupScenarioContext
```

Comment:
I guess your goal is to solve these TODO's in another PR?

Reply:
@bergquist Yes, it's a larger operation (as part of dropping goconvey), and I think best done separately.